editor: Fix block cursor offset when selecting text

lennartkloock opened the pull request:
Closes #36677 and #20121

### Problem

Vim visual mode and Helix selection mode both require the cursor to be on the last character of the selection. Until now, this was implemented by offsetting the cursor one character to the left whenever a block cursor is used. (Since the visual modes use a block cursor.)
https://github.com/zed-industries/zed/blob/1683052e6ce44d206e947c397917b2ab7617c0fe/crates/editor/src/element.rs#L153
However, this oversees the problem that **some users might want to use the block cursor without being in visual mode**. Meaning that the cursor is offset by one character to the left even though Vim/Helix mode isn't even activated.

### Solution

Since the Vim mode implementation is separate from the `editor` crate the solution is not as straightforward as just checking the current vim mode. Therefore this PR introduces a new `Editor` struct field called `cursor_offset_on_selection`. This field replaces the condition above and is set to `true` whenever the Vim mode is changed to a visual mode, and `false` otherwise.

If there is a better solution without having to add a new `Editor` field, let me know.

Release Notes:

- Fixes block and hollow cursor being offset when selecting text

lennartkloock committed:
editor: Fix block cursor offset on select

lennartkloock committed:
editor: Fix test

lennartkloock commented:
Unfortunately, the `editor::vim_flavor` function doesn't help here because we have to figure out not only if the editor is in Vim mode but also if it is in **visual** mode.

That said, what we could do is use `editor::vim_flavor` to check if it is in Vim mode and then use the `cursor_shape` field to determine if it is in visual mode but in my opinion that field shouldn't be used like that because it results in bugs like this one.

dinocosta commented:
@aurexav Thank you for sharing those steps! When you state "1. Use block cursor", do yo mean the following settings?

```json
{
  "cursor_shape": "block",
}
```

If so, I just tested these changes, comparing how it behaves against the latest stable release and it appears the behavior for `shift-end` is still the same ▼

https://github.com/user-attachments/assets/a9f0d1bf-05ca-4927-aed0-5320d0720b5d

The behavior of `editor::SelectToEndOfLine` does appear to be broken but it seems that the changes in this Pull Request don't actually fix it? Let me know if I'm missing something, thanks!

> And you can continue press `right`. And the cursor won't move to next line. The selection is completely wrong.

If you have vim mode enabled, I believe this is meant to replicate vim's behavior, where if you're at the end of the line, pressing `l` will not continue the selection on the next line.

lennartkloock commented:
> @aurexav Thank you for sharing those steps! When you state "1. Use block cursor", do yo mean the following settings?
> 
> ```json
> {
>   "cursor_shape": "block",
> }
> ```
> 
> If so, I just tested these changes, comparing how it behaves against the latest stable release and it appears the behavior for `shift-end` is still the same ▼
> CleanShot.2025-12-05.at.10.28.57.mp4
> 
> The behavior of `editor::SelectToEndOfLine` does appear to be broken but it seems that the changes in this Pull Request don't actually fix it? Let me know if I'm missing something, thanks!
> 
> > And you can continue press `right`. And the cursor won't move to next line. The selection is completely wrong.
> 
> If you have vim mode enabled, I believe this is meant to replicate vim's behavior, where if you're at the end of the line, pressing `l` will not continue the selection on the next line.

I don't know if there is a misunderstanding here but it seems like you are still in Vim mode in your video. The issue that is fixed here only appears if you're NOT in Vim mode. Please disable Vim mode for the reproduction steps to work. (set `"vim_mode": false` in your settings file)

lennartkloock commented:
Here is a video I recorded demonstrating the issue. I hope that helps to further clarify it.

Settings are:
```json
{
  "vim_mode": false,
  "cursor_shape": "block",
}
```

https://github.com/user-attachments/assets/d296cb27-b719-43f4-820c-37c0931268fd

dinocosta committed:
refactor: move editor update to sync_vim_settings

Since we're already updating the vim's editor in that function, we can
avoid an extre call to `self.update_editor`.
